### ISSUE TITLE
Gdk-CRITICAL error

### ISSUE THREAD

**Issue by zensubz** [bleachbit/bleachbit#919]
Hi,

I use a custom bleachbit.ini file kept at /root/.config/bleachbit, and run bleachbit in CLI mode (bleachbit --preset -c) to clean unneeded files. I have noticed that the current version (v4.0.x, python3-based) produces the following error message if I run bleachbit (in CLI mode) inside a chroot environment:

(bleachbit:32577): Gdk-CRITICAL **: 10:45:13.712: gdk_screen_get_root_window: assertion 'GDK_IS_SCREEN (screen)' failed

Although this doesn't affect the cleaning process (inside chroot), I am noticing this error only after upgrading to v4.0.x. Is there any fix for this..?

Thanks.

**Comment by AmanAjayVarma** [comment]
I also have this problem

```bash
No protocol specified
Unable to init server: Could not connect: Connection refused
No protocol specified
Unable to init server: Could not connect: Connection refused

(bleachbit:27825): Gdk-CRITICAL **: 00:33:16.301: gdk_screen_get_root_window: assertion 'GDK_IS_SCREEN (screen)' failed
No protocol specified
Unable to init server: Could not connect: Connection refused

(bleachbit:27825): Gtk-WARNING **: 00:33:16.310: cannot open display: :0
```

**Comment by AmanAjayVarma** [comment]
I solved this problem with going to `/etc/ssh/sshd_config` and entering
```
X11Forwarding yes
```
Or 

1. Installing `sudo apt install putty`
2. Opening putty
3. Connection -> SHH -> X11
4. - [ ] Enable X11 forwarding

**Comment by zensubz** [comment]
In my case, it has already been uncommented in the default sshd_config...

**Comment by AmanAjayVarma** [comment]
Then try to uncomment `X11UseForwarding yes` also.

**Comment by ediowar** [comment]
@AmanAjayVarma Which your OS?

**Comment by AmanAjayVarma** [comment]
Linux Debian with LXQt and i3wm.

**Comment by ediowar** [comment]
@AmanAjayVarma 

> Then try to uncomment `X11UseForwarding yes` also.

`X11UseForwarding yes` is not in openSUSE Tumbleweed `/etc/ssh/sshd_config`
[sshd_config.txt](https://github.com/bleachbit/bleachbit/files/4675570/sshd_config.txt)

**Comment by abitrolly** [comment]
@az0 I'll work on this one as I am on Fedora too.

**Comment by abitrolly** [comment]
@zensubz what is your OS and desktop environment? You can run `bleachbit --sysinfo` to get the info and paste the output here. The similar issue on Fedora 33 in #981 is not reproducible anymore. Maybe this one is also fixed.

@FillyAgioro look like you run BleachBit in SSH session? Does the problem repeat without SSH?

**Comment by FillyAgioro** [comment]
@abitrolly I don't ran bleachbit in SSH session. But try to install Bleachbit from website maybe that would work. The apt version is very old. The new version on website is new and works fine for me.

**Comment by abitrolly** [comment]
Waiting for @zensubz reply to decide how to move on with this issue.

**Comment by zensubz** [comment]
@abitrolly the OS I have observed this issue is CentOS 7.8, and as such, the official repo does not provide a recent version of bleachbit. The 'Gdk-CRITICAL error' I have mentioned occurs only when I run bleachbit inside a chroot environment - for example, I make a remaster of CentOS using livecd-creator (which is python2 based), and run bleachbit (inside chroot) as an additional step just before the live-img-data gets compressed. The earlier version of bleachbit I had used (2.x, python2 based) didn't encounter this issue, whereas the more recent versions of bleachbit (4.x, python3 based) throws this error just before starting the process (scan/cleanup); it does clean the junk files inside chroot. Thanks for the follow-up.

**Comment by abitrolly** [comment]
We can use `docker` to test if BleachBit in CLI mode still tries to access GUI subsystem.

**Comment by az0** [comment]
> We can use `docker` to test if BleachBit in CLI mode still tries to access GUI subsystem.

How about just disabling the GUI like this?

```
unset XAUTHORITY
unset DISPLAY
python3 bleachbit.py -p google_chrome.cookies
```

With latest Git revision and Ubuntu 20.04 I see this

```
Unable to init server: Could not connect: Connection refused
Unable to init server: Could not connect: Connection refused

(bleachbit.py:518063): Gdk-CRITICAL **: 08:31:57.131: gdk_screen_get_root_window: assertion 'GDK_IS_SCREEN (screen)' failed
```
It is normal for the GUI to be needed to clean system.clipboard, but I think that's the only option that requires it.

The original comment was about CLI mode inside chroot, so I modified the issue title to reflect this. In  other words, the function is fine, but we want to avoid an error message. There was another comment about the same error with sudo. Especially if the goal is to run the GUI with sudo, the issue could be better tracked in its own issue ticket.

**Comment by huastone2018** [comment]
uncomment in  (it is alread uncomment when i to check, not modified by me, os is ubuntu 18.4.5 LTS), but it doesn't work. see below:

#AllowAgentForwarding yes
#AllowTcpForwarding yes
#GatewayPorts no
X11Forwarding yes
#X11DisplayOffset 10
#X11UseLocalhost yes

**Comment by abitrolly** [comment]
I can finally repeat that even with `-h` with latest code. Looking into what causes Gtk import.
```
# unset XAUTHORITY
# unset DISPLAY
# ./bleachbit.py -h
Unable to init server: Could not connect: Connection refused
Unable to init server: Could not connect: Connection refused

(bleachbit.py:387705): Gdk-CRITICAL **: 19:26:05.171: gdk_screen_get_root_window: assertion 'GDK_IS_SCREEN (screen)' failed
No work to do. Specify options.
```

**Comment by abitrolly** [comment]
Interesting that unsetting `XAUTHORITY` and `DISPLAY` in user session has no effect. It only repeats in chroot.

**Comment by abitrolly** [comment]
Another tricky thing about this bug. It doesn't manifest itself if `PyGObject` is not installed, like for example, in `virtualenv`.

**Comment by abitrolly** [comment]
I am using `python-hunter` to trace call flow.
```
PYTHONHUNTER="~Q(filename_contains='venv'),stdlib=False,action=CallPrinter(stream=sys.stdout,force_colors=True)" python ./bleachbit.py -h
```
Now that I've reproduced the error in `virtualenv`, I just need to redirect the stderr to catch Gdk message together with the trace at that moment.

**Comment by abitrolly** [comment]
Here we go.
```bash
$ PYTHONHUNTER="~Q(filename_contains='venv'),stdlib=False,action=CallPrinter(force_colors=True)" python ./bleachbit.py -h &> 1
$ less -R 1
```
```python
[...]ot/p/bleachbit/bleachbit/Cleaner.py:37    line               import warnings
[...]ot/p/bleachbit/bleachbit/Cleaner.py:38    line               warnings.simplefilter("ignore", Warning)
[...]ot/p/bleachbit/bleachbit/Cleaner.py:39    line               try:
[...]ot/p/bleachbit/bleachbit/Cleaner.py:40    line               from bleachbit.GuiBasic import Gtk, Gdk
[...]t/p/bleachbit/bleachbit/GuiBasic.py:20    call               => <module>()
[...]t/p/bleachbit/bleachbit/GuiBasic.py:20    line                  """
[...]t/p/bleachbit/bleachbit/GuiBasic.py:24    line                  from bleachbit import _
[...]t/p/bleachbit/bleachbit/GuiBasic.py:26    line                  import os
[...]t/p/bleachbit/bleachbit/GuiBasic.py:28    line                  import gi
[...]t/p/bleachbit/bleachbit/GuiBasic.py:29    line                  gi.require_version('Gtk', '3.0')
[...]t/p/bleachbit/bleachbit/GuiBasic.py:30    line                  from gi.repository import Gtk, Gdk  # keep after gi.require_version()
Unable to init server: Could not connect: Connection refused
Unable to init server: Could not connect: Connection refused
[...]t/p/bleachbit/bleachbit/GuiBasic.py:32    line                  if os.name == 'nt':
[...]t/p/bleachbit/bleachbit/GuiBasic.py:36    line                  def browse_folder(parent, title, multiple, stock_button):
[...]t/p/bleachbit/bleachbit/GuiBasic.py:65    line                  def browse_file(parent, title):
[...]t/p/bleachbit/bleachbit/GuiBasic.py:89    line                  def browse_files(parent, title):
[...]t/p/bleachbit/bleachbit/GuiBasic.py:114   line                  def delete_confirmation_dialog(parent, mention_preview, shred_settings=False):
[...]t/p/bleachbit/bleachbit/GuiBasic.py:153   line                  def message_dialog(parent, msg, mtype=Gtk.MessageType.ERROR, buttons=Gtk.ButtonsType.OK):
[...]t/p/bleachbit/bleachbit/GuiBasic.py:168   line                  def open_url(url, parent_window=None, prompt=True):
[...]t/p/bleachbit/bleachbit/GuiBasic.py:168   return             <= <module>: None
[...]ot/p/bleachbit/bleachbit/Cleaner.py:41    line               HAVE_GTK = Gdk.get_default_root_window() is not None

(bleachbit.py:399980): Gdk-CRITICAL **: 20:53:58.131: gdk_screen_get_root_window: assertion 'GDK_IS_SCREEN (screen)' failed
[...]ot/p/bleachbit/bleachbit/Cleaner.py:49    line               if 'posix' == os.name:
[...]ot/p/bleachbit/bleachbit/Cleaner.py:50    line               from bleachbit import Unix
[...]iroot/p/bleachbit/bleachbit/Unix.py:22    call               => <module>()
[...]iroot/p/bleachbit/bleachbit/Unix.py:22    line                  """
[...]iroot/p/bleachbit/bleachbit/Unix.py:26    line                  import bleachbit
[...]iroot/p/bleachbit/bleachbit/Unix.py:27    line                  from bleachbit import FileUtilities, General
[...]iroot/p/bleachbit/bleachbit/Unix.py:28    line                  from bleachbit import _
```

**Comment by abitrolly** [comment]
The only usage of Gtk and Gdk API here are `Gtk.RecentManager` and `Gtk.Clipboard`. 
```
# ag "(Gtk|Gdk)" bleachbit/Cleaner.py
40:    from bleachbit.GuiBasic import Gtk, Gdk
41:    HAVE_GTK = Gdk.get_default_root_window() is not None
477:                Gtk.RecentManager().purge_items()
553:                clipboard = Gtk.Clipboard.get(Gdk.SELECTION_CLIPBOARD)
```

**Comment by abitrolly** [comment]
Regardless of if the code needs display or not, importing Gtk shows the error. Now looking into PyGObject.
```
In [1]: import gi
   ...: gi.require_version("Gtk", "3.0")
   ...: from gi.repository import Gtk
Unable to init server: Could not connect: Connection refused
Unable to init server: Could not connect: Connection refused
```

**Comment by abitrolly** [comment]
I opened an issue with upstream about how to suppress the "Unable to init server" message, or delay the initialization until it is needed - https://gitlab.gnome.org/GNOME/pygobject/-/issues/453 Looks like `RecentManager()` doesn't need this connection to work.
```python
In [1]: import gi
   ...: gi.require_version('Gtk', '3.0')
   ...: from gi.repository.Gtk import RecentManager
Unable to init server: Could not connect: Connection refused
Unable to init server: Could not connect: Connection refused

In [2]: RecentManager().get_items()
Out[2]: [<Gtk.RecentInfo object at 0x7f0b020e6880 (GtkRecentInfo at 0x55ee4122bc30)>]
```

**Comment by abitrolly** [comment]
The error comes from the line `initialized, argv = Gtk.init_check(sys.argv)` found in https://gitlab.gnome.org/GNOME/pygobject/-/blob/master/gi/overrides/Gtk.py

```bash
# cat testimport.py
import gi
gi.require_version('Gtk', '3.0')
from gi.repository.Gtk import RecentManager
# PYTHONHUNTER="stdlib=False,action=CallPrinter(force_colors=True)" python testimport.py &> 2
# less -R 2
```
```python
[...]9/site-packages/gi/overrides/Gtk.py:1654  line                  __all__.append('stock_lookup')
[...]9/site-packages/gi/overrides/Gtk.py:1656  line                  if Gtk._version == "4.0":
[...]9/site-packages/gi/overrides/Gtk.py:1659  line                  initialized, argv = Gtk.init_check(sys.argv)
[...]ython3.9/site-packages/gi/module.py:120   call                  => __getattr__(self=<gi.module.IntrospectionModule object at 0x7f49ee28b880>, name
='init_check')
[...]ython3.9/site-packages/gi/module.py:121   line                     info = repository.find_by_name(self._namespace, name)
[...]ython3.9/site-packages/gi/module.py:122   line                     if not info:
[...]ython3.9/site-packages/gi/module.py:126   line                     if isinstance(info, EnumInfo):
[...]ython3.9/site-packages/gi/module.py:162   line                     elif isinstance(info, RegisteredTypeInfo):
[...]ython3.9/site-packages/gi/module.py:213   line                     elif isinstance(info, FunctionInfo):
[...]ython3.9/site-packages/gi/module.py:214   line                     wrapper = info
[...]ython3.9/site-packages/gi/module.py:223   line                     self.__dict__[name] = wrapper
[...]ython3.9/site-packages/gi/module.py:224   line                     return wrapper
[...]ython3.9/site-packages/gi/module.py:224   return                <= __getattr__: <gi.FunctionInfo object at 0x7f49ebfa9830>
Unable to init server: Could not connect: Connection refused
[...]9/site-packages/gi/overrides/Gtk.py:1660  line                  sys.argv = list(argv)
[...]9/site-packages/gi/overrides/Gtk.py:1660  return             <= <module>: None
```

`PyGObject` does this `init_check` before returning `Gtk` object. There is quite an old issue about skipping the initialization - https://gitlab.gnome.org/GNOME/pygobject/-/issues/72

**Comment by az0** [comment]
@abitrolly 
Thanks for troubleshooting and for filing the ticket

**Comment by zilexa** [comment]
Hi have this issue unfortunately, with version 4.4.2 on Manjaro Gnome. 

bleachbit --sysinfo
```
BleachBit version 4.4.2
GTK version 3.24.36
GTK theme = Adw-dark
GTK icon theme = Papirus-Dark-Maia
GTK prefer dark theme = False
SQLite version 3.40.1
local_cleaners_dir = None
locale_dir = /usr/share/locale/
options_dir = /root/.config/bleachbit
personal_cleaners_dir = /root/.config/bleachbit/cleaners
system_cleaners_dir = /usr/share/bleachbit/cleaners
locale.getdefaultlocale = ('en_GB', 'UTF-8')
os.getenv('DESKTOP_SESSION') = None
os.getenv('LOGNAME') = root
os.getenv('USER') = root
os.getenv('SUDO_UID') = 1000
os.path.expanduser('~') = /root
platform.linux_distribution() = ('ManjaroLinux', '22.0.4', 'Sikaris')
platform.platform = Linux-5.15.94-1-MANJARO-x86_64-with-glibc2.37
platform.version = #1 SMP PREEMPT Tue Feb 14 19:45:58 UTC 2023
sys.argv = ['/usr/bin/bleachbit', '--sysinfo']
sys.executable = /usr/bin/python3
sys.version = 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0]
__file__ = /usr/share/bleachbit/SystemInformation.py
```

**Comment by az0** [comment]
Thank you for your patience. This fell through the cracks. The latest stable release is BleachBit 5.0.2, which in earlier versions may have fixed or improved this. Still, I made additional changes today that help.

Please test [BleachBit 5.0.2 (stable)](https://www.bleachbit.org/news/bleachbit-502) or the latest changes from Git (shown below).

```sh
# Create a temporary directory in the home folder
mkdir -p ~/bleachbit-testing
cd ~/bleachbit-testing

# Clean up any previous clone
rm -rf bleachbit

# Clone the repository
git clone https://github.com/bleachbit/bleachbit.git
cd bleachbit

# Switch to the dev branch with the changed discussed here
git checkout -b dev origin/dev

# Run the test
python3 bleachbit.py --sysinfo
```

Did that help?

Cross reference: #1486
